src/client/client.cpp

```diff
@@ -1792,18 +1792,78 @@ void Client::makeScreenshot()
 	if (fs::IsPathAbsolute(g_settings->get("screenshot_path")))
 		screenshot_dir = g_settings->get("screenshot_path");
 	else
-		screenshot_dir = porting::path_user + DIR_DELIM + g_settings->get("screenshot_path");
+		screenshot_dir = porting::path_user + std::string(DIR_DELIM) + g_settings->get("screenshot_path");
+
+	// Using Irrlicht methods since those defined in filesys.cpp are
+	// not reliable, they do not return correct absolute path in some
+	// path values such as the root or path with non-existing folders
+	io::IFileSystem *irrfs = RenderingEngine::get_filesystem();
+	io::path path = irrfs->getAbsolutePath(screenshot_dir.c_str());
+	std::string screenshot_dir_abs = std::string(path.c_str());
+
+	std::size_t pos = screenshot_dir_abs.rfind(std::string(DIR_DELIM));
+	if (screenshot_dir_abs.length() > 1 && pos == screenshot_dir_abs.length()-1)
+		screenshot_dir_abs = screenshot_dir_abs.substr(0, pos);
+
+	if (screenshot_dir_abs != "/") {
+		bool mtpath_done = false, screenpath_done = false, levels_diverged=false;
+		int mtpath_levels = 0, common_levels = 0;
+		std::size_t mtpath_pos, screenpath_pos;
+		std::string mtpath_str = fs::AbsolutePath(porting::path_user), screenpath_str = screenshot_dir_abs;
+		std::string mtpath_name, screenpath_name, common_path;
+
+		do {
+			if (!mtpath_done) mtpath_levels++;
+			mtpath_pos = mtpath_str.find(std::string(DIR_DELIM));
+			if (mtpath_pos != std::string::npos) {
+				mtpath_name = mtpath_str.substr(0, mtpath_pos);
+				mtpath_str.erase(0, mtpath_pos + std::string(DIR_DELIM).length());
+			} else if (!mtpath_done) {
+				mtpath_name = mtpath_str;
+				mtpath_done = true;
+			}
+			screenpath_pos = screenpath_str.find(std::string(DIR_DELIM));
+			if (screenpath_pos != std::string::npos) {
+				screenpath_name = screenpath_str.substr(0, screenpath_pos);
+				screenpath_str.erase(0, screenpath_pos + std::string(DIR_DELIM).length());
+			} else if (!screenpath_done) {
+				screenpath_name = screenpath_str;
+				screenpath_done = true;
+			}
+			if (!levels_diverged && mtpath_name == screenpath_name) {
+				common_path += mtpath_name
+					+ ((!mtpath_done && !screenpath_done) ? std::string(DIR_DELIM) : "");
+				common_levels++;
+			}
+			else if (!levels_diverged) levels_diverged = true;
+		} while ((mtpath_pos != std::string::npos) || (screenpath_pos != std::string::npos));
```

Comment:
The parentheses around the `!=` are redundant.

Reply:
done.